include/zephyr/drivers/retained_mem.h

```diff
@@ -0,0 +1,196 @@
+ * non-initialised RAM region, or in specific registers, but is not reset when a different
+ * application begins execution or the device is rebooted (without power loss). It must support
+ * byte-level reading and writing without a need to erase data before writing.
+ *
+ * Note that drivers must implement all functions, none of the functions are optional.
+ */
+struct retained_mem_driver_api {
+	retained_mem_size_api size;
+	retained_mem_read_api read;
+	retained_mem_write_api write;
+	retained_mem_clear_api clear;
+};
+
+/**
+ * @brief		Returns the size of the retained memory area.
+ *
+ * @param dev		Retained memory device to use.
+ *
+ * @retval		Positive value indicating size in bytes on success, else negative errno
+ *			code.
+ */
+__syscall ssize_t retained_mem_size(const struct device *dev);
+
+static inline ssize_t z_impl_retained_mem_size(const struct device *dev)
+{
+	struct retained_mem_driver_api *api = (struct retained_mem_driver_api *)dev->api;
+
+	return api->size(dev);
+}
+
+/**
+ * @brief		Reads data from the Retained memory area.
+ *
+ * @param dev		Retained memory device to use.
+ * @param offset	Offset to read data from.
+ * @param buffer	Buffer to store read data in.
+ * @param size		Size of data to read.
+ *
+ * @retval		0 on success else negative errno code.
+ */
+__syscall int retained_mem_read(const struct device *dev, off_t offset, uint8_t *buffer,
+				size_t size);
+
+static inline int z_impl_retained_mem_read(const struct device *dev, off_t offset,
+					   uint8_t *buffer, size_t size)
+{
+	struct retained_mem_driver_api *api = (struct retained_mem_driver_api *)dev->api;
+	size_t area_size;
+
+	/* Validate user-supplied parameters */
+	if (size == 0) {
+		return 0;
+	} else if (offset < 0) {
+		return -EINVAL;
+	}
+
+	area_size = api->size(dev);
+
+	if (size > area_size || (area_size - size) < (size_t)offset) {
+		return -EINVAL;
+	}
+
+	return api->read(dev, offset, buffer, size);
+}
+
+/**
+ * @brief		Writes data to the Retained memory area - underlying data does not need to
+ *			be cleared prior to writing.
+ *
+ * @param dev		Retained memory device to use.
+ * @param offset	Offset to write data to.
+ * @param buffer	Data to write.
+ * @param size		Size of data to be written.
+ *
+ * @retval		0 on success else negative errno code.
+ */
+__syscall int retained_mem_write(const struct device *dev, off_t offset, const uint8_t *buffer,
+				 size_t size);
+
+static inline int z_impl_retained_mem_write(const struct device *dev, off_t offset,
+					    const uint8_t *buffer, size_t size)
+{
+	struct retained_mem_driver_api *api = (struct retained_mem_driver_api *)dev->api;
+	size_t area_size;
+
+	/* Validate user-supplied parameters */
+	if (size == 0) {
+		return 0;
+	} else if (offset < 0) {
+		return -EINVAL;
+	}
+
+	area_size = api->size(dev);
+
+	if (size > area_size || (area_size - size) < (size_t)offset) {
+		return -EINVAL;
+	}
```

Comment:
Move this up and concatenate with `offset < 0` check.

Reply:
Done